### DISCUSSION TITLE
PostgREST timeouts when loading schema cache in multi-tenant environment with large user base and schemas

### DISCUSSION THREAD

**Discussion by akshaysasidrn** [PostgREST/postgrest#3774] [Q&A] [answered]
I am facing an issue where PostgREST times out when attempting to load the schema cache in a multi-tenant PostgreSQL environment with approximately 100k schemas. Each schema has a dedicated PostgreSQL user, and PostgREST is configured to impersonate the corresponding user when querying its schema. The timeout occurs around 60 seconds after the request begins I think.

I have tried changing up the timeouts with `SET statement_timeout = '5min';` with no luck

### Environment:

- PostgreSQL Version: 13.16
- PostgREST Version: 12.2.0
- Configuration:
  - PGRST_DB_URI: postgres://postgres:xxxxxx@<DB_HOST>:5432/tooljet_db
  - PGRST_DB_ANON_ROLE: postgres (added to debug better)
  - PGRST_DB_POOL_ACQUISITION_TIMEOUT: 120s
  - PGRST_DB_PRE_CONFIG: postgrest.pre_config
 
 ```sql
CREATE OR REPLACE FUNCTION postgrest.pre_config()
RETURNS void AS $$
  SELECT
    set_config('pgrst.db_aggregates_enabled', 'true', false);
  SELECT
    set_config('pgrst.db_schemas', string_agg(nspname, ','), true)
  FROM pg_namespace
  WHERE nspname LIKE 'workspace_%';
$$ LANGUAGE sql;

```

PostgREST logs:
```
08/Nov/2024:12:51:21 +0000: Starting PostgREST 12.2.0 (ec89f6b)...
08/Nov/2024:12:51:21 +0000: Admin server listening on port 3002
08/Nov/2024:12:51:21 +0000: Listening on port 3000
08/Nov/2024:12:51:21 +0000: Attempting to connect to the database...
08/Nov/2024:12:51:53 +0000: Successfully connected to PostgreSQL 13.16 on x86_64-pc-linux-gnu, compiled by Debian clang version 12.0.1, 64-bit
08/Nov/2024:12:52:24 +0000: Listening for notifications on the "pgrst" channel
192.168.65.1 - user_8d18eead-eac3-4f57-ba66-4ec368bfaccf [08/Nov/2024:12:56:44 +0000] "GET /954f82c0-2796-4b04-b520-16deb36a7714?&limit=50&offset=0 HTTP/1.1" 503 - "" ""
192.168.65.1 - user_8d18eead-eac3-4f57-ba66-4ec368bfaccf [08/Nov/2024:12:56:44 +0000] "GET /954f82c0-2796-4b04-b520-16deb36a7714?&limit=50&offset=0 HTTP/1.1" 503 - "" ""
08/Nov/2024:12:57:27 +0000: {"code":"57014","details":null,"hint":null,"message":"canceling statement due to statement timeout"}
08/Nov/2024:12:57:27 +0000: Failed to query the role settings. {"code":"57014","details":null,"hint":null,"message":"canceling statement due to statement timeout"}
08/Nov/2024:12:57:27 +0000: Config reloaded
08/Nov/2024:12:59:30 +0000: {"code":"57014","details":null,"hint":null,"message":"canceling statement due to statement timeout"}
08/Nov/2024:12:59:30 +0000: Failed to load the schema cache. {"code":"57014","details":null,"hint":null,"message":"canceling statement due to statement timeout"}
```

Are there recommendations for managing statement timeouts and schema configurations in such a large multi-tenant architecture?

**Comment by wolfgangwalther** [comment]
> I have tried changing up the timeouts with SET statement_timeout = '5min'; with no luck

How did you change that?

**Comment by wolfgangwalther** [comment]
> PGRST_DB_ANON_ROLE: postgres

So you are impersonating 100k users, but use the SUPERUSER `postgres` as the anonymous role?

I really hope this is just your reproducer and not actually happening :D

**Reply by akshaysasidrn** [reply]
> I really hope this is just your reproducer and not actually happening :D

In the actual production setup, I do not use the postgres SUPERUSER as the ANON role. This configuration is only for testing and debugging purposes on my local PostgREST setup. :D

> How did you change that?

I connected to the database with psql and ran the statement. Restarted my PostgREST server after that.

**Reply by wolfgangwalther** [reply]
> and ran the statement.

Which statement exactly?

Did you do `ALTER ROLE authenticator SET ...`?

**Reply by akshaysasidrn** [reply]
I connected to the database with the SUPERUSER `postgres`, given that I had kept the ANON role the same.
Then I ran `SET statement_timeout = '5min'` which as I understand would affect the user in session. Tried restarting PostgREST after that thinking it would pick up ANON role.

> Did you do ALTER ROLE authenticator SET ...?
No, I haven't done this.

**Reply by wolfgangwalther** [reply]
Just `SET` will only change your current session. But you need to change the session where PostgREST runs in. Thus, you need to change the default settings for the role that you are using as the authenticator role. If that's `postgres` in your test setup.. then you need to do `ALTER ROLE postgres SET statement_timeout =...;`. Then this will be in effect for the next session of that user as well. Then PostgREST should pick it up at startup.

**Reply by akshaysasidrn** [reply]
Thank you for the explanation! Giving it a try. 

Could you provide an overview of what happens behind the scenes during these processes or help point me to a resource so that I can better understand? 

And, are there any optimization strategies you would suggest for such a setup? For example, if a user is allowed to create or modify tables within a PostgreSQL schema, I would need to trigger a notification to reload the schema. What would be the best way to optimize for such cases if these modifications happen frequently? Any guidance would be greatly appreciated!

**Reply by wolfgangwalther** [reply]
I don't think PostgREST in it's current form is designed to handle 100k different schemas, especially considering reloads and how long it will take to load the schema cache.

You should run **many** PostgREST instances, imho.

But maybe @steve-chavez has more experience in that direction.

**Reply by steve-chavez** [reply]
I've also had a case with multi-tenant dynamic schemas and saw terribly slow schema cache loads (user reported it loaded after 2 hours). This dynamic schemas had a pattern like `client-<uuid>`. The user was open to changing the multi-tenancy design to single schema with RLS though, so I didn't pursue a solution.

@akshaysasidrn Do your dynamic schemas have a naming pattern?

I'm wondering if we should just add configuration for exclusion of these dynamic schemas, like:

```ini
db-excluded-schemas-pattern = 'client-%'
```

**Reply by wolfgangwalther** [reply]
Would that help? The way I understand it it's the opposite here: There are 100k *exposed* schemas, so they all need to be loaded?

**Reply by akshaysasidrn** [reply]
@steve-chavez Yes, I do have a naming pattern like `workspace_%`.  This is set through a `pre_config()` function.

```sql
CREATE OR REPLACE FUNCTION postgrest.pre_config()
RETURNS void AS $$
 SELECT
   set_config('pgrst.db_aggregates_enabled', 'true', false);
 SELECT
   set_config('pgrst.db_schemas', string_agg(nspname, ','), true)
 FROM pg_namespace
 WHERE nspname LIKE 'workspace_%';
$$ LANGUAGE sql;
```
In my case, I don't think I can exclude the schemas pattern as I let the users create tables and do CRUD on respective schemas.
Is a single schema with RLS the only way to go with PostgREST there? I will explore that approach then. I wanted to do a separation of tenants within individual schemas for better isolation and security.

**Reply by steve-chavez** [reply, marked as answer]
> There are 100k exposed schemas, so they all need to be loaded?

Right, that wouldn't help.

> Is a single schema with RLS the only way to go with PostgREST there? I will explore that approach then.

Yes, that's the only way for now.

Also  we're now logging the schema cache load time https://github.com/PostgREST/postgrest/pull/3779.
